File: c/parallel/src/solver_SR_3.c
/****
	SOLVER_SEND_RECV
***/
#include <stdio.h>
#include <math.h>
#include "mpi.h"
#include "precision.h"
#include "allocate.h"
static MPI_Status  *sta1,*sta2;
static MPI_Request *req1,*req2;
static KINT NFLAG=0;
extern FILE *fp_log;
void SOLVER_SEND_RECV_3( int N, int NEIBPETOT,
	int NEIBPE[], int IMPORT_INDEX[], int IMPORT_ITEM[],
	int EXPORT_INDEX[], int EXPORT_ITEM[],
	KREAL WS[], KREAL WR[], KREAL X[], int my_rank)
{

	int  ii,k,neib,istart,inum;
/***
	INIT.
***/
	if( NFLAG == 0 ){
		sta1=(MPI_Status*)allocate_vector(sizeof(MPI_Status),NEIBPETOT);
		sta2=(MPI_Status*)allocate_vector(sizeof(MPI_Status),NEIBPETOT);
		req1=(MPI_Request*)allocate_vector(sizeof(MPI_Request),NEIBPETOT);
		req2=(MPI_Request*)allocate_vector(sizeof(MPI_Request),NEIBPETOT);
		NFLAG=1;
	} 
/***
	SEND
***/
	for( neib=1;neib<=NEIBPETOT;neib++){
		istart=EXPORT_INDEX[neib-1];
		inum  =EXPORT_INDEX[neib]-istart;
		for( k=istart+1;k<=istart+inum;k++){
			ii=3*EXPORT_ITEM[k-1];
			WS[3*k-3]=X[ii-3];
			WS[3*k-2]=X[ii-2];
			WS[3*k-1]=X[ii-1];
		}

		MPI_Isend(&WS[3*istart],3*inum,MPI_DOUBLE,
				  NEIBPE[neib-1],0,MPI_COMM_WORLD,&req1[neib-1]);
	}
/***
	RECEIVE
***/
	for( neib=1;neib<=NEIBPETOT;neib++){
		istart=IMPORT_INDEX[neib-1];
		inum  =IMPORT_INDEX[neib]-istart;
		MPI_Irecv(&WR[3*istart],3*inum,MPI_DOUBLE,
				  NEIBPE[neib-1],0,MPI_COMM_WORLD,&req2[neib-1]);
	}

	MPI_Waitall (NEIBPETOT, req2, sta2);

	for( neib=1;neib<=NEIBPETOT;neib++){
		istart=IMPORT_INDEX[neib-1];
		inum  =IMPORT_INDEX[neib]-istart;
		for( k=istart+1;k<=istart+inum;k++){
			ii   = 3*IMPORT_ITEM[k-1];
			X[ii-3]=WR[3*k-3];
			X[ii-2]=WR[3*k-2];
			X[ii-1]=WR[3*k-1];
		}
	}

	MPI_Waitall (NEIBPETOT, req1, sta1);

}

File: f/parallel/src/solver_SR_3.f
!C
!C*** 
!C*** module solver_SR_3
!C***
!C
      module solver_SR_3
      contains
!C
!C*** SOLVER_SEND_RECV
!C
      subroutine  SOLVER_SEND_RECV_3                                    &
     &                ( N, NEIBPETOT, NEIBPE, IMPORT_INDEX, IMPORT_ITEM,&
     &                                        EXPORT_INDEX, EXPORT_ITEM,&
     &                  WS, WR, X, my_rank)

      implicit REAL*8 (A-H,O-Z)
      include  'mpif.h'
      include  'precision.inc'

      integer(kind=kint )                , intent(in)   ::  N
      integer(kind=kint )                , intent(in)   ::  NEIBPETOT
      integer(kind=kint ), pointer :: NEIBPE      (:)
      integer(kind=kint ), pointer :: IMPORT_INDEX(:)
      integer(kind=kint ), pointer :: IMPORT_ITEM  (:)
      integer(kind=kint ), pointer :: EXPORT_INDEX(:)
      integer(kind=kint ), pointer :: EXPORT_ITEM  (:)
      real   (kind=kreal), dimension(3*N), intent(inout):: WS
      real   (kind=kreal), dimension(3*N), intent(inout):: WR
      real   (kind=kreal), dimension(3*N), intent(inout):: X
      integer                            , intent(in)   :: my_rank

      integer(kind=kint ), dimension(:,:), save, allocatable :: sta1
      integer(kind=kint ), dimension(:,:), save, allocatable :: sta2
      integer(kind=kint ), dimension(:  ), save, allocatable :: req1
      integer(kind=kint ), dimension(:  ), save, allocatable :: req2  

      integer(kind=kint ), save :: NFLAG
      data NFLAG/0/

!C
!C-- INIT.
      if (NFLAG.eq.0) then
        allocate (sta1(MPI_STATUS_SIZE,NEIBPETOT))
        allocate (sta2(MPI_STATUS_SIZE,NEIBPETOT))
        allocate (req1(NEIBPETOT))
        allocate (req2(NEIBPETOT))
        NFLAG= 1
      endif
       
!C
!C-- SEND
      do neib= 1, NEIBPETOT
        istart= EXPORT_INDEX(neib-1)
        inum  = EXPORT_INDEX(neib  ) - istart
        do k= istart+1, istart+inum
               ii   = 3*EXPORT_ITEM(k)
           WS(3*k-2)= X(ii-2)
           WS(3*k-1)= X(ii-1)
           WS(3*k  )= X(ii  )
        enddo

        call MPI_ISEND (WS(3*istart+1), 3*inum,MPI_DOUBLE_PRECISION,    &
     &                  NEIBPE(neib), 0, MPI_COMM_WORLD, req1(neib),    &
     &                  ierr)
      enddo

!C
!C-- RECEIVE
      do neib= 1, NEIBPETOT
        istart= IMPORT_INDEX(neib-1)
        inum  = IMPORT_INDEX(neib  ) - istart
        call MPI_IRECV (WR(3*istart+1), 3*inum, MPI_DOUBLE_PRECISION,   &
     &                  NEIBPE(neib), 0, MPI_COMM_WORLD, req2(neib),    &
     &                  ierr)
      enddo

      call MPI_WAITALL (NEIBPETOT, req2, sta2, ierr)
   
      do neib= 1, NEIBPETOT
        istart= IMPORT_INDEX(neib-1)
        inum  = IMPORT_INDEX(neib  ) - istart
        do k= istart+1, istart+inum
            ii   = 3*IMPORT_ITEM(k)
          X(ii-2)= WR(3*k-2)
          X(ii-1)= WR(3*k-1)
          X(ii  )= WR(3*k  )
        enddo
      enddo

      call MPI_WAITALL (NEIBPETOT, req1, sta1, ierr)

      end subroutine solver_send_recv_3
      end module     solver_SR_3
